Add user-guide for clone API (#564)
kubevirt/user-guide: docs/operations/clone_api.md

> # Clone API

Reviewer: Is clone aso in Alpha? If so, please add a disclaimer

Author: done

## Prerequesites

### Snapshot / Restore

Under the hood, the clone API relies upon Snapshot & Restore APIs. Therefore, in order to be able to use the clone API,
please see [Snapshot & Restore prerequesites](./snapshot_restore_api.md/#prerequesites).

### Snapshot Feature Gate

Currently, clone API is gaurded by Snapshot feature gate. The
[feature gates](./activating_feature_gates.md#how-to-activate-a-feature-gate)
field in the KubeVirt CR must be expanded by adding the `Snapshot` to it.

## The clone object

Firstly, as written above, the clone API relies upon Snapshot & Restore APIs under the hood. Therefore, it might be helpful
to look at [Snapshot & Restore](./snapshot_restore_api.md) user-guide page for more info.

### VirtualMachineClone object overview

In order to initiate clone, a `VirtualMachineClone` object (CRD) needs to be created on the cluster. An example
for such an object is:
```yaml
kind: VirtualMachineClone
apiVersion: "clone.kubevirt.io/v1alpha1"
metadata:
  name: testclone

spec:
  # source & target definitions
  source:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: vm-cirros

>   target:

Reviewer: Not relevant fr the user guide, but: Was it considered to call it destination?

>   target:

Reviewer: Was it considered to call it `from:` and `to:`?

    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: vm-clone-target

  # labels & annotations definitions
  labelFilters:
    - "*"
    - "!someKey/*"
  annotationFilters:
    - "anotherKey/*"

  # other identity stripping specs:
  newMacAddresses:
    interfaceName: "00-11-22"
  newSMBiosSerial: "new-serial"
```

In the next section I will go through the different settings to elaborate them.

#### Source & Target

The source and target indicate the source/target API group, kind and name. A few important notes:
* Currently, the only supported kind is `VirtualMachine`, but more types are expected to be supported in the future.
  See "future roadmap" below for more info.

> * The target name is **optional**. If unspecified, the clone controller will generate a name for the target automatically.

Reviewer: In the same naamespace I guess?

Author: Yes, currently only same-namespace-clones are supported, I'll add that

#### Label & Annotation filters

These spec fields are intended to determine which labels / annotations are being copied to the target or stripped away.

The filters are a list of strings. Each string represents a key that may exist at the source. Every source key that matches
to one of these values is being copied to the cloned target. In addition, special regular-expression-like characters can be
used:

* **Wildcard character (\*)** can be used to match to anything. Wildcard can be only used at the **end** of the filter.
  * These filters are valid:
    * "*"
    * "some/key*"
  * These filters are invalid:
    * "some/*/key"
    * "*/key"
* **Negation character (!)** can be used to avoid matching certain keys. Negation can be only used at the **beginning** of a filter.
  Note that a Negation and Wildcard can be used together.
  * These filters are valid:
    * "!some/key"
    * "!some/*"
  * These filters are invalid:
    * "key!"
    * "some/!key"

Setting label / annotation filters is **optional**. If unset, all labels / annotations will be copied as a default.

> #### newMacAddresses

Reviewer: Can we add a seciton called `Handling of identity` or some thing similar.

WHat happens during cloning to
- mac (covered below)
- smbios (covered below)
- disk serial numbers
- any volume (dv, pvc, secret, configmap, …)

Author: Ehum..
It's already covered under the description of the spec fields. I'm feeling like it would simply duplicate the information.

Regarding disk serial / volume cloning - this is being done with snapshot / restore which ultimately uses VolumeSnapshot/VolumeRestore, their implementation is different depending on your CSI driver.

This field is used to explicitly replace MAC addresses for certain interfaces. The field is a string to string map; the
keys represent interface names and the values represent the new MAC address for the clone target.

This field is optional. By default, all mac addresses are stripped out. This suits situations when kube-mac-pool is
deployed in the cluster which would automatically assign the target with a fresh valid MAC address.

#### newSMBiosSerial

This field is used to explicitly set an SMBios serial for the target.

This field is optional. By default, the target would have an empty serial.

### Creating a VirtualMachineClone object

After the clone manifest is ready, we can create it:
```bash

> kubectl create -f clone.yaml

Reviewer: Is there actually work to implement

`virtctl clone <vmname> [<optionalnewname>]`?

Author: Not yet, but there sure should be!

```

To wait for a clone to complete, execute:
```bash

> kubectl wait vmclone testclone --for condition=Ready

Reviewer: Wohoo!

field in the KubeVirt CR must be expanded by adding the `Snapshot` to it.

## The clone object

Firstly, as written above, the clone API relies upon Snapshot & Restore APIs under the hood. Therefore, it might be helpful
to look at [Snapshot & Restore](./snapshot_restore_api.md) user-guide page for more info.

### VirtualMachineClone object overview

In order to initiate clone, a `VirtualMachineClone` object (CRD) needs to be created on the cluster. An example
for such an object is:
```yaml
kind: VirtualMachineClone
apiVersion: "clone.kubevirt.io/v1alpha1"
metadata:
  name: testclone

spec:
  # source & target definitions
  source:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: vm-cirros
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: vm-clone-target

  # labels & annotations definitions
  labelFilters:
    - "*"
    - "!someKey/*"
  annotationFilters:
    - "anotherKey/*"

  # other identity stripping specs:
  newMacAddresses:
    interfaceName: "00-11-22"
  newSMBiosSerial: "new-serial"
```

In the next section I will go through the different settings to elaborate them.

#### Source & Target

> The source and target indicate the source/target API group, kind and name. A few important notes:

Reviewer 2: These do not get displayed properly

* Currently, the only supported kind is `VirtualMachine`, but more types are expected to be supported in the future.
  See "future roadmap" below for more info.
* The target name is **optional**. If unspecified, the clone controller will generate a name for the target automatically.
* The target and source must reside in the same namespace.

#### Label & Annotation filters

These spec fields are intended to determine which labels / annotations are being copied to the target or stripped away.

The filters are a list of strings. Each string represents a key that may exist at the source. Every source key that matches
to one of these values is being copied to the cloned target. In addition, special regular-expression-like characters can be
used:

> * **Wildcard character (\*)** can be used to match to anything. Wildcard can be only used at the **end** of the filter.

Reviewer 2: ```suggestion
* **Wildcard character (\*)** can be used to match anything. Wildcard can be only used at the **end** of the filter.
```

  * These filters are valid:
    * "*"
    * "some/key*"
  * These filters are invalid:
    * "some/*/key"
    * "*/key"
* **Negation character (!)** can be used to avoid matching certain keys. Negation can be only used at the **beginning** of a filter.
  Note that a Negation and Wildcard can be used together.
  * These filters are valid:
    * "!some/key"
    * "!some/*"
  * These filters are invalid:
    * "key!"
    * "some/!key"

Setting label / annotation filters is **optional**. If unset, all labels / annotations will be copied as a default.

#### newMacAddresses

This field is used to explicitly replace MAC addresses for certain interfaces. The field is a string to string map; the
keys represent interface names and the values represent the new MAC address for the clone target.

This field is optional. By default, all mac addresses are stripped out. This suits situations when kube-mac-pool is
deployed in the cluster which would automatically assign the target with a fresh valid MAC address.

#### newSMBiosSerial

This field is used to explicitly set an SMBios serial for the target.

> This field is optional. By default, the target would have an empty serial.

Reviewer 2: In that case, I believe the serial will be generated from the name of the VM and won't be a different random value every time you boot the VM. Could be worth saying something about that. "An empty serial" sounds scary, like the VM will have no SMBios serial, or the serial will keep changing, which are both highly undesirable.

Author: Thanks! You're absolutely right. Changed to:
```
By default, the target would have an auto-generated serial that's based on the VM name.
```

### Creating a VirtualMachineClone object

After the clone manifest is ready, we can create it:
```bash
kubectl create -f clone.yaml
```

To wait for a clone to complete, execute:
```bash
kubectl wait vmclone testclone --for condition=Ready
```

> You can check the clone's phase in the clone's status. It can be one of the following:

Reviewer 2: Rendering issue

* Progressing
* Succeeded

After the clone is finished, the target can be inspected:
```bash
kubectl get vm vm-clone-target -o yaml
```

## Future roadmap

> The clone API is in an early alpha version and may be changed dramatically. There are many improvements and features

Reviewer 2: ```suggestion
The clone API is in an early alpha version and may change dramatically. There are many improvements and features
```

that are expected to be added, the most significant goals are:

* Add more supported source types. The most important is `VirtualMachineSnapshot` (see section below for more info). We
  also intend to support `VirtualMachineInstace` types in the future.
* Add a cross-namespace clone support. This needs to be supported for snapshots / restores first.

### Using clones as a "golden VM image"

> One of the great things that could be accomplished with the clone API when `VirthalMachineSnapshot`s are supported

Reviewer 3: ```suggestion
One of the great things that could be accomplished with the clone API when `VirtualMachineSnapshot`s are supported
```

Author: Thanks! Fixed

as source types is to create "golden VM images" (a.k.a. Templates / Bookmark VMs / etc). In other words, the following
workflow would be available:

> **Create a golden image**

Reviewer 2: Rendering issue
